src/hoc/withTheme.js

```diff
@@ -33,45 +39,59 @@ const wrapWithTheme = (Component: ReactClass<any>) => {
       [CHANNEL_NEXT]: CONTEXT_CHANNEL_SHAPE,
     }
 
-    state: { theme?: ?Object } = {}
+    static getDerivedStateFromProps(
+      nextProps: {
+        theme?: ?Object,
+        [key: string]: any,
+      },
+      prevState: State
+    ) {
+      const theme = determineTheme(
+        nextProps,
+        prevState.theme,
+        prevState.defaultProps
```

Comment:
`defaultProps` are static, so you should be good to obtain them from `WithTheme.defaultProps` – this way you'll at least avoid weird constructs like putting defaultProps in state.

Reply:
Makes sense, thanks

Reply:
Unfortunately, I can't do the same for `Styled(Native)Component`, because inheritance is used there:

https://github.com/JetBrains/styled-components/blob/fb2e97f8e08fb660a4453cdc79040d133a2bfe1b/src/models/StyledComponent.js#L317

So should I do it for `withTheme` only, or keep it as is for consistency?

Reply:
I'd still do it for WithTheme, even when it's troublesome (but probably doable) to get it working in other places

Reply:
Maybe I could just move `gDSFP` from `BaseComponent` to `StyledComponent`, will try later today

Reply:
Inheritance 😭😭😭

Reply:
We can also just get rid of `BaseComponent` for now. It's a perf optimisation for v8, but it's probably not worth the mess with this refactor anymore. I'll think about a babel transpilation that can do this automatically maybe. Either way, it's not worth keeping if that means that it complicates this PR and leads to worse code.

Reply:
Plus, there is a `withComponent` case, I'm not sure what would be a better soulution to handle it

Reply:
We wouldn't recreate the methods in `Object.assign`, only create new references to them

Reply:
> Plus, there are withComponent and extend case, I'm not sure what would be a better solution to handle those

This is a more tough part. In this case `ParentComponent` is not `BaseComponent` but rather the one being extended

Reply:
As far as I remember, extend doesn’t imply an actual inheritance. We just create a new config object based on the previous (parent) component and concatenate all rules.

If we are actually applying inheritance then I don’t see it being actually necessary

Reply:
`Object.assign(StyledComponent, Base)` is still inheritance though, just written differently.

Reply:
@kitten OK, so which changes are you suggesting now?